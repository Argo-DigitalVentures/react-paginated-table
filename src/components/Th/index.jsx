import React, { PropTypes } from 'react';
import find from 'lodash.find';
import get from 'lodash.get';

import {
  dataType,
  columnType,
  classNameType,
  styleType,
  sortType,
} from '../../propTypes';

import { evaluate } from '../../utils';

function getSortDirection(columnId, sorts) {
  const sort = find(sorts, sort => sort.id === columnId);
  return get(sort, 'direction', null);
}

function Th(props) {
  const {
    style: cellStyle,
    className: cellClassName,
    column,
    columnIndex,
    columns,
    data,
    sorts,
    handleHeaderClick,
    ...rest
  } = props;
  const sortDirection = getSortDirection(column.id, sorts);
  const dataProps = { column, columnIndex, columns, data, sortDirection };
  const style = evaluate(cellStyle, dataProps);
  const className = evaluate(cellClassName, dataProps);
  return (
    <th // eslint-disable-line jsx-a11y/no-static-element-interactions
      style={style}
      className={className}
      onClick={e => handleHeaderClick(column.id, e.shiftKey)}
      {...rest}
    >
      {
        typeof column.header === 'function' ? (
          <column.header {...dataProps} />
        ) : column.header
      }
    </th>
  );
}

Th.defaultProps = {
  style: {},
  className: '',
};

Th.propTypes = {
  column: columnType.isRequired,
  columnIndex: PropTypes.number.isRequired,
  columns: PropTypes.arrayOf(columnType).isRequired,
  data: dataType.isRequired,
  sorts: PropTypes.arrayOf(sortType).isRequired,
  handleHeaderClick: PropTypes.func.isRequired,
  style: styleType,
  className: classNameType,
};

export default Th;
