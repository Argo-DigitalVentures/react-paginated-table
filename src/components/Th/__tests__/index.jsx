import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import Th from '../';

const data = [
  { company: 'Argo', location: 'New York' },
  { company: 'John Hancock', location: 'New York' },
];

const columns = [
  { id: 'company', header: 'Company Name', selector: 'company' },
  { id: 'location', header: () => null, selector: 'location' },
];

describe('<Th />', () => {
  it('should accept className, style, and other DOM props', () => {
    const tree = shallow(
      <Th
        style={{
          color: 'green',
        }}
        className="test"
        column={columns[0]}
        columnIndex={0}
        columns={columns}
        data={data}
        sorts={[]}
        handleHeaderClick={() => null}
      />,
    );
    expect(toJson(tree)).toMatchSnapshot();
  });

  it('should accept className and style functions', () => {
    const styleFn = jest.fn().mockReturnValue({
      color: 'green',
    });
    const classNameFn = jest.fn().mockReturnValue('test');
    const tree = shallow(
      <Th
        style={styleFn}
        className={classNameFn}
        column={columns[0]}
        columnIndex={0}
        columns={columns}
        data={data}
        sorts={[]}
        handleHeaderClick={() => null}
      />,
    );
    expect(toJson(tree)).toMatchSnapshot();
    expect(styleFn).toBeCalledWith({
      column: columns[0],
      columnIndex: 0,
      columns,
      data,
      sortDirection: null,
    });
    expect(classNameFn).toBeCalledWith({
      column: columns[0],
      columnIndex: 0,
      columns,
      data,
      sortDirection: null,
    });
  });


  it('should use a custom header renderer if available', () => {
    const tree = shallow(
      <Th
        style={{
          color: 'green',
        }}
        className="test"
        column={columns[1]}
        columnIndex={1}
        columns={columns}
        data={data}
        sorts={[]}
        handleHeaderClick={() => null}
      />,
    );
    expect(toJson(tree)).toMatchSnapshot();
  });

  it('should pass the sort direction to the style and className functions', () => {
    const classNameFn = jest.fn().mockReturnValue('test');
    const styleFn = jest.fn().mockReturnValue({
      color: 'green',
    });
    shallow(
      <Th
        style={styleFn}
        className={classNameFn}
        column={columns[0]}
        columnIndex={0}
        columns={columns}
        data={data}
        sorts={[{
          id: columns[0].id,
          direction: 'ASC',
        }]}
        handleHeaderClick={() => null}
      />,
    );
    expect(styleFn).toBeCalledWith({
      column: columns[0],
      columnIndex: 0,
      columns,
      data,
      sortDirection: 'ASC',
    });
    expect(classNameFn).toBeCalledWith({
      column: columns[0],
      columnIndex: 0,
      columns,
      data,
      sortDirection: 'ASC',
    });
  });

  it('should call the handleHeaderClick prop with column id and event shiftKey property', () => {
    const handleHeaderClickSpy = jest.fn();
    const tree = shallow(
      <Th
        column={columns[0]}
        columnIndex={0}
        columns={columns}
        data={data}
        sorts={[]}
        handleHeaderClick={handleHeaderClickSpy}
      />,
    );
    tree.find('th').simulate('click', { shiftKey: false });
    expect(handleHeaderClickSpy).toBeCalledWith(columns[0].id, false);
    tree.find('th').simulate('click', { shiftKey: true });
    expect(handleHeaderClickSpy).toBeCalledWith(columns[0].id, false);
  });
});
