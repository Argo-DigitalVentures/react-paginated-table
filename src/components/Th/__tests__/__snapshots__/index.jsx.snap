exports[`<Th /> should accept className and style functions 1`] = `
<th
  className="test"
  onClick={[Function]}
  style={
    Object {
      "color": "green",
    }
  }>
  Company Name
</th>
`;

exports[`<Th /> should accept className, style, and other DOM props 1`] = `
<th
  className="test"
  onClick={[Function]}
  style={
    Object {
      "color": "green",
    }
  }>
  Company Name
</th>
`;

exports[`<Th /> should use a custom header renderer if available 1`] = `
<th
  className="test"
  onClick={[Function]}
  style={
    Object {
      "color": "green",
    }
  }>
  <header
    column={
      Object {
        "header": [Function],
        "id": "location",
        "selector": "location",
      }
    }
    columnIndex={1}
    columns={
      Array [
        Object {
          "header": "Company Name",
          "id": "company",
          "selector": "company",
        },
        Object {
          "header": [Function],
          "id": "location",
          "selector": "location",
        },
      ]
    }
    data={
      Array [
        Object {
          "company": "Argo",
          "location": "New York",
        },
        Object {
          "company": "John Hancock",
          "location": "New York",
        },
      ]
    }
    sortDirection={null} />
</th>
`;
